### ISSUE TITLE
Sending a file from client to server

### ISSUE THREAD

**Issue by timijntema** [oktal/pistache#517]
Currently I am working on a project where I need to send image files from one Docker container to another. Pistache seems like a great option for client - server communcation. However, I cannot find anything about sending a file from the client to the server using this library. I have found the `Http::serveFile` function, but it looks like that is only meant to send a file from the server to a client. I have found the FileBuffer and StreamBuf as well, but I cannot figure out how to send these buffers unless I would try to create a Response and send that from the client.

Can anybody tell me how to do this? What would a send file function look like? Is this even possible?

**Comment by timijntema** [comment]
This question seems to have been asked in #166 as well.

**Comment by timijntema** [comment]
Another thing to mention is that I have tried to send the file as the body by reading the bytes as characters and then sending the compiled string. This however crashes the server with a segmentation fault, I assume due to the buffer being to small.

**Comment by timijntema** [comment]
I now tested it with a much larger buffer size as well. No difference. Still a segmentation fault. The buffersize is set larger than the size of the image.

**Comment by janhkrueger** [comment]
From my understanding you want to send a file from server a to server b.
Not sure if Pistache is really the way to go for this.

Have you tried sending the file via ssh / scp from one container to the other?
Is the file actively send from a or is it requested by b?

**Comment by dennisjenkins75** [comment]
timijntema: Write the simplest HTTP server using pistache that you can that should serve a file (say "/etc/motd" or something).  Test it with wget or curl.  Post the source here, or link to the source.

Compile it with "-O0 -ggdb".  Run it under gdb.  When it crashed, get the stack of the crashing thread ("bt") and post that.

**Comment by timijntema** [comment]
Thanks for the quick response. I noticed after reading through all the comments of #263 that I was setting the maximum buffer size incorrectly. The code below is my server side. Even after adding the buffer size change to the options, I cannot send 4096 characters. The message I receive is `Request exceeded maximum buffer size`. I can however send less characters without any problem using the exact same client code (1024 characters is no problem).

I managed to find the exact amount at which it fails. 3954 characters is fine, 3955 characters generates the buffer message. Interesting to note is that the entire server crashes as soon as i go anywhere above the 3955 limit.

I build Pistache from sources previous week (not sure exactly what commit).
```
struct HelloHandler : public Pistache::Http::Handler {
  HTTP_PROTOTYPE(HelloHandler)
  void onRequest(const Pistache::Http::Request & request, Pistache::Http::ResponseWriter writer) override{
      writer.send(Pistache::Http::Code::Ok, "Hello, World!");
  }
};

int main(int argc, char ** argv) {
    const std::string host = "*";
    std::shared_ptr<Pistache::Http::Endpoint> server = std::make_shared<Pistache::Http::Endpoint>(Pistache::Address(host, port));
    auto options = Pistache::Http::Endpoint::options();
    options.threads(1);
    options.flags(Pistache::Tcp::Options::InstallSignalHandler);
    options.maxPayload(4096);
       
    server->init(options);
    server->setHandler(Pistache::Http::make_handler<HelloHandler>());
    server->serve();
    server->shutdown();
    return 0;
}
```

**Comment by timijntema** [comment]
Retested with the exact client code example provided in the repository and then making two changes. I added:
```
std::string data = "";
for (int i = 0; i < 3995; i++) {
    data += "j";
}
```
And I made the send line send a body instead of a cookie.
`auto resp = client.get(page).body(data).send()`

And strangely enough it was able to suddenly send 3995 characters. However, as soon as I changed the send line to:
`auto resp = client.get(page).header<Pistache::Http::Header::ContentType>(Pistache::Http::Mime::MediaType(Pistache::Http::Mime::Type::Image, Pistache::Http::Mime::Subtype::Jpeg)).body(data).send();`

The error occured again. The reason it kept on failing is most likely due to the header line adding more bytes to the request which made it go over the buffer limit. This is supported by it working again when I changed the limit to 6000.

I am sorry for wasting your time. Thank you for the quick responses.

One last think I would like to note. Maybe some of these issues that are not realy issues could be avoided with some extra documentation. Adding Doxygen comments to functions might help for understanding their usage for instance. I am noticing that I spent most of my time trying to find what functions to use and then having to test every single function I find to see what they do. Its just a suggestion.
